node/src/components/console.rs

```diff
@@ -0,0 +1,163 @@
+//! Unix console component.
+//!
+//! The console listens on a configurable unix socket for incoming connections and allows deep debug
+//! access to a running node via special commands.
+
+use std::{
+    fmt::{self, Display, Formatter},
+    fs, io,
+    path::{Path, PathBuf},
+};
+
+use datasize::DataSize;
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+use tokio::{net::UnixListener, sync::watch};
+use tracing::debug;
+
+use super::Component;
+use crate::{
+    effect::{console::DumpConsensusStateRequest, EffectBuilder, EffectExt, Effects},
+    reactor::EventQueueHandle,
+    types::NodeRng,
+    utils::umask,
+    WithDir,
+};
+use util::ShowUnixAddr;
+
+mod command;
+mod tasks;
+mod util;
+
+/// Unix console component.
+#[derive(Debug, DataSize)]
+pub(crate) struct Console {
+    /// Sender, when dropped, will cause server and client connections to exit.
+    #[data_size(skip)]
+    #[allow(dead_code)] // only used for its `Drop` impl.
+    shutdown_sender: watch::Sender<()>,
+}
+
+/// Unix console configuration.
+#[derive(Clone, DataSize, Debug, Deserialize)]
+pub(crate) struct Config {
+    /// Whether or not the console is enabled.
+    enabled: bool,
+    /// Path to listen on.
+    socket_path: PathBuf,
+    /// `umask` to apply before creating the socket.
+    socket_umask: u16,
+}
+
+impl Default for Config {
+    fn default() -> Self {
+        Self {
+            enabled: false,
+            socket_path: "debug.socket".into(),
+            socket_umask: 0o077,
+        }
+    }
+}
+
+impl Console {
+    /// Creates a new Unix console component.
+    pub(crate) fn new<REv>(
+        cfg: &WithDir<Config>,
+        event_queue: EventQueueHandle<REv>,
+    ) -> Result<(Self, Effects<Event>), Error>
+    where
+        REv: From<DumpConsensusStateRequest> + Send,
+    {
+        let config = cfg.value();
+        let (shutdown_sender, shutdown_receiver) = watch::channel(());
```

Comment:
Pardon me my incompetence, but how does this work? We're creating a channel here but we never use `sender` part. Is the design here that `shutdown_sender` will be dropped as the `Console` component gets dropped in the `joiner`/`participating` reactor which will then send a signal to `shutdown_receiver` that the corresponding `sender` was dropped? Which in turn will close the socket connection?

Reply:
That's exactly right :)

Reply:
Awesome.